{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9338d2d-9fa4-4403-9362-eb74675a975b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Script Name: ChainLAMP_coverage_analysis_v3.4_extension.ipynb\n",
    "Version: v3.4\n",
    "Author: Shane Gilligan-Steinberg\n",
    "Date: 240822\n",
    "\n",
    "Description:\n",
    "\n",
    "Additional software: Script Name: ChainLAMP_coverage_analysis_v3.4_extension.ipynb\n",
    "This can be used to generate a sets of input sequences organized by subtype for use in pipeline\n",
    "[1] Generate library of sequences to be inputted into the pipeline (from LANL alignment - need to remove gaps and organize by subtype).\n",
    "Other capability of splitting by year is not available in this version\n",
    "    [1B] Generate alignments with additional split by year of sequeneces\n",
    "    [1C] Another option is to gather sequences from GenBank IDs\n",
    "\"\"\"\n",
    "\n",
    "# Install all neccesary libraries\n",
    "!pip install biopython pandas matplotlib tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6524f880-9018-4bc8-846c-e12877f2caa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "[1] Generate library of sequences to be inputted into the pipeline (from LANL alignment - need to remove gaps and organize by subtype).\n",
    "\n",
    "Author: Shane Gilligan-Steinberg\n",
    "Date: 240715\n",
    "\n",
    "Description:\n",
    "Create subtype files of HIV subtypes from LANL alignments\n",
    "Distinction from LANL sequence naming criteria.\n",
    "\"\"\"\n",
    "\n",
    "import os\n",
    "import csv\n",
    "\n",
    "def remove_gaps_from_sequences(input_file, min_length=0):\n",
    "    \"\"\"\n",
    "    Remove gaps ('-') from sequences in a FASTA file, return a dictionary of sequences,\n",
    "    and optionally omit sequences below a minimum length.\n",
    "    Args:\n",
    "        input_file (str): Path to the input FASTA file.\n",
    "        min_length (int): Minimum length of sequences to be retained.\n",
    "    Returns:\n",
    "        Dict[str, List[str]]: Dictionary with sequence names as keys and lists of cleaned sequences as values.\n",
    "    \"\"\"\n",
    "    fasta_records = {}\n",
    "    current_name = None\n",
    "    current_seq = []\n",
    "    original_counts = {}\n",
    "    \n",
    "    with open(input_file, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('>'):\n",
    "                if current_name is not None:\n",
    "                    # Process the previous record\n",
    "                    cleaned_seq = ''.join(current_seq).replace('-', '')\n",
    "                    if current_name not in original_counts:\n",
    "                        original_counts[current_name] = 0\n",
    "                    original_counts[current_name] += 1\n",
    "                    if len(cleaned_seq) >= min_length:\n",
    "                        if cleaned_seq:\n",
    "                            if current_name not in fasta_records:\n",
    "                                fasta_records[current_name] = []\n",
    "                            if cleaned_seq not in fasta_records[current_name]:\n",
    "                                fasta_records[current_name].append(cleaned_seq)\n",
    "                \n",
    "                # Start a new record\n",
    "                current_name = line.strip()[1:]\n",
    "                current_seq = []\n",
    "            else:\n",
    "                current_seq.append(line.strip())\n",
    "        \n",
    "        # Process the last record\n",
    "        if current_name is not None:\n",
    "            cleaned_seq = ''.join(current_seq).replace('-', '')\n",
    "            if current_name not in original_counts:\n",
    "                original_counts[current_name] = 0\n",
    "            original_counts[current_name] += 1\n",
    "            if len(cleaned_seq) >= min_length:\n",
    "                if cleaned_seq:\n",
    "                    if current_name not in fasta_records:\n",
    "                        fasta_records[current_name] = []\n",
    "                    if cleaned_seq not in fasta_records[current_name]:\n",
    "                        fasta_records[current_name].append(cleaned_seq)\n",
    "    \n",
    "    return fasta_records, original_counts\n",
    "\n",
    "def split_fasta_by_name(input_file, csv_output_file, output_directory, min_length=0):\n",
    "    \"\"\"\n",
    "    Split a FASTA file into separate files based on sequence name prefixes and create a CSV file with output file names.\n",
    "    Args:\n",
    "        input_file (str): Path to the input FASTA file.\n",
    "        csv_output_file (str): Path to the CSV output file.\n",
    "        output_directory (str): Directory where the output files will be saved.\n",
    "        min_length (int): Minimum length of sequences to be retained.\n",
    "    \"\"\"\n",
    "    possible_names = ['A', 'B', 'C', 'D', 'F', 'G', 'H', 'J', 'K', 'L', '01_AE', '02_AG']  # List of possible names\n",
    "    \n",
    "    fasta_records, original_counts = remove_gaps_from_sequences(input_file, min_length)\n",
    "    \n",
    "    output_files = []\n",
    "    alignment_counts = {}\n",
    "    \n",
    "    if not os.path.exists(output_directory):\n",
    "        os.makedirs(output_directory)\n",
    "    \n",
    "    # Adjust sequence names and write to different output files based on name prefix\n",
    "    for name, seqs in fasta_records.items():\n",
    "        adjusted_name = name.split('.')[0]\n",
    "        if adjusted_name[0] == 'A' or adjusted_name[0] == 'F':\n",
    "            adjusted_name = adjusted_name[0]\n",
    "        \n",
    "        output_file = os.path.join(output_directory, f\"{adjusted_name}.fasta\" if adjusted_name in possible_names else \"other.fasta\")\n",
    "        \n",
    "        with open(output_file, 'a') as out_file:\n",
    "            for seq in seqs:\n",
    "                out_file.write(f\">{name}\\n{seq}\\n\")\n",
    "        \n",
    "        if output_file not in output_files:\n",
    "            output_files.append(output_file)\n",
    "            alignment_counts[output_file] = 0\n",
    "        \n",
    "        alignment_counts[output_file] += len(seqs)\n",
    "    \n",
    "    # Write the output file names and counts to the specified CSV file\n",
    "    with open(csv_output_file, 'w', newline='') as csvfile:\n",
    "        csv_writer = csv.writer(csvfile)\n",
    "        csv_writer.writerow(['subtypes', 'original_alignments', 'cleaned_alignments'])  # Write column header\n",
    "        for output_file in output_files:\n",
    "            subtype = os.path.basename(output_file).split('.')[0]\n",
    "            original_count = sum(original_counts.get(name, 0) for name in original_counts if name.startswith(subtype))\n",
    "            csv_writer.writerow([os.path.join(\"FLT_MCRF_LTR_500-850\", os.path.basename(output_file)), original_count, alignment_counts[output_file]])\n",
    "    \n",
    "    # Print counts\n",
    "    print(\"FASTA sequences have been split into different output files based on name prefix and cleaned of '-' characters.\")\n",
    "    print(f\"Output file names and alignment counts have been written to {csv_output_file}.\")\n",
    "    print(\"Alignment counts per file:\")\n",
    "    for output_file in output_files:\n",
    "        subtype = os.path.basename(output_file).split('.')[0]\n",
    "        original_count = sum(original_counts.get(name, 0) for name in original_counts if name.startswith(subtype))\n",
    "        print(f\"{os.path.join('FLT_MCRF_LTR_500-850', os.path.basename(output_file))}: original alignments = {original_count}, cleaned alignments = {alignment_counts[output_file]}\")\n",
    "\n",
    "# Example usage:\n",
    "input_file = \"Alignments/FLT_MCRF_LTR_500-850/HIV1_FLT_2021_500-850_DNA.fasta\"\n",
    "csv_output_file = \"Subtypes_LTR_500-850.csv\"\n",
    "output_directory = \"Alignments/FLT_MCRF_LTR_500-850\"\n",
    "min_length = 300  # Example minimum length for sequences to be retained\n",
    "split_fasta_by_name(input_file, csv_output_file, output_directory, min_length)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d93187f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "[1B] Generate library of sequences to be inputted into the pipeline (from LANL alignment - need to remove gaps and organize by subtype). Enables a split by year\n",
    "\n",
    "Author: Shane Gilligan-Steinberg\n",
    "Date: 240715\n",
    "\n",
    "Description:\n",
    "Create subtype files of HIV subtypes from LANL alignments\n",
    "Distinction from LANL sequence naming criteria.\n",
    "Also split files by year into 5 year bins\n",
    "\"\"\"\n",
    "\n",
    "import os\n",
    "import csv\n",
    "from collections import defaultdict\n",
    "\n",
    "def remove_gaps_from_sequences(input_file):\n",
    "    \"\"\"\n",
    "    Remove gaps ('-') from sequences in a FASTA file and return a dictionary of sequences.\n",
    "    Args:\n",
    "        input_file (str): Path to the input FASTA file.\n",
    "    Returns:\n",
    "        Dict[str, List[str]]: Dictionary with sequence names as keys and lists of cleaned sequences as values.\n",
    "    \"\"\"\n",
    "    fasta_records = {}\n",
    "    current_name = None\n",
    "    current_seq = []\n",
    "    \n",
    "    with open(input_file, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('>'):\n",
    "                if current_name is not None:\n",
    "                    # Process the previous record\n",
    "                    cleaned_seq = ''.join(current_seq).replace('-', '')\n",
    "                    if cleaned_seq:\n",
    "                        if current_name not in fasta_records:\n",
    "                            fasta_records[current_name] = []\n",
    "                        if cleaned_seq not in fasta_records[current_name]:\n",
    "                            fasta_records[current_name].append(cleaned_seq)\n",
    "                \n",
    "                # Start a new record\n",
    "                current_name = line.strip()[1:]\n",
    "                current_seq = []\n",
    "            else:\n",
    "                current_seq.append(line.strip())\n",
    "        \n",
    "        # Process the last record\n",
    "        if current_name is not None:\n",
    "            cleaned_seq = ''.join(current_seq).replace('-', '')\n",
    "            if cleaned_seq:\n",
    "                if current_name not in fasta_records:\n",
    "                    fasta_records[current_name] = []\n",
    "                if cleaned_seq not in fasta_records[current_name]:\n",
    "                    fasta_records[current_name].append(cleaned_seq)\n",
    "    \n",
    "    return fasta_records\n",
    "\n",
    "def get_year_bin(year):\n",
    "    \"\"\"\n",
    "    Get the bin for the given year.\n",
    "    Args:\n",
    "        year (int): Year to be binned.\n",
    "    Returns:\n",
    "        str: The bin name.\n",
    "    \"\"\"\n",
    "    if 90 <= year <= 99:\n",
    "        return \"90+99\"\n",
    "    elif 0 <= year <= 4:\n",
    "        return \"00+04\"\n",
    "    elif 5 <= year <= 9:\n",
    "        return \"05+09\"\n",
    "    elif 10 <= year <= 14:\n",
    "        return \"10+14\"\n",
    "    elif 15 <= year <= 25:\n",
    "        return \"15+25\"\n",
    "    else:\n",
    "        return \"other\"\n",
    "\n",
    "def split_fasta_by_year_and_subtype(input_file, csv_output_file, output_directory):\n",
    "    \"\"\"\n",
    "    Split a FASTA file into separate files based on the year in the sequence name and create a CSV file with output file names.\n",
    "    Args:\n",
    "        input_file (str): Path to the input FASTA file.\n",
    "        csv_output_file (str): Path to the CSV output file.\n",
    "        output_directory (str): Directory where the output files will be saved.\n",
    "    \"\"\"\n",
    "    possible_names = ['A', 'B', 'C', 'D', 'G', '01_AE', '02_AG']  # List of possible names\n",
    "    \n",
    "    fasta_records = remove_gaps_from_sequences(input_file)\n",
    "    \n",
    "    output_files = []\n",
    "    alignment_counts = defaultdict(int)\n",
    "    \n",
    "    if not os.path.exists(output_directory):\n",
    "        os.makedirs(output_directory)\n",
    "    \n",
    "    # Adjust sequence names and write to different output files based on year and subtype\n",
    "    for name, seqs in fasta_records.items():\n",
    "        year_part = name.split('.')[2]  # Assuming the year is the third part of the name separated by dots\n",
    "        if year_part.isdigit():  # Check if the year part is numeric\n",
    "            year_bin = get_year_bin(int(year_part))\n",
    "        else:\n",
    "            year_bin = \"other\"\n",
    "        \n",
    "        subtype = name.split('.')[0]\n",
    "        if subtype not in possible_names:\n",
    "            if subtype[0] == 'A': # or subtype[0] == 'F':\n",
    "                subtype = subtype[0]\n",
    "            else:\n",
    "                subtype = \"other\"\n",
    "        \n",
    "        output_file = os.path.join(output_directory, f\"{year_bin}_{subtype}.fasta\")\n",
    "        \n",
    "        with open(output_file, 'a') as out_file:\n",
    "            for seq in seqs:\n",
    "                out_file.write(f\">{name}\\n{seq}\\n\")\n",
    "        \n",
    "        if output_file not in output_files:\n",
    "            output_files.append(output_file)\n",
    "        \n",
    "        alignment_counts[(year_bin, subtype)] += len(seqs)\n",
    "    \n",
    "    # Write the output file names and counts to the specified CSV file\n",
    "    with open(csv_output_file, 'w', newline='') as csvfile:\n",
    "        csv_writer = csv.writer(csvfile)\n",
    "        csv_writer.writerow(['year_bin', 'subtype', 'total_sequences'])  # Write column header\n",
    "        for output_file in output_files:\n",
    "            basename = os.path.basename(output_file).split('.')[0]\n",
    "            year_bin, subtype = basename.split('_', 1)  # Split only on the first underscore\n",
    "            csv_writer.writerow([year_bin, subtype, alignment_counts[(year_bin, subtype)]])\n",
    "    \n",
    "    # Print counts\n",
    "    print(\"FASTA sequences have been split into different output files based on year and subtype, and cleaned of '-' characters.\")\n",
    "    print(f\"Output file names and sequence counts have been written to {csv_output_file}.\")\n",
    "    print(\"Sequence counts per file:\")\n",
    "    for output_file in output_files:\n",
    "        basename = os.path.basename(output_file).split('.')[0]\n",
    "        year_bin, subtype = basename.split('_', 1)  # Split only on the first underscore\n",
    "        print(f\"{os.path.join(output_directory, os.path.basename(output_file))}: total sequences = {alignment_counts[(year_bin, subtype)]}\")\n",
    "\n",
    "# Example usage:\n",
    "input_file = \"Alignments/FLT_MCRF_Pol/HIV1_FLT_2021_pol_DNA.fasta\"\n",
    "csv_output_file = \"Alignments/FLT_MCRF_Pol/Yearly/Yearly_Subtypes_Pol.csv\"\n",
    "output_directory = \"Alignments/FLT_MCRF_Pol/Yearly\"\n",
    "split_fasta_by_year_and_subtype(input_file, csv_output_file, output_directory)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ece810a8-b025-4e0b-a409-026d0f97be00",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "[1C] Another option is to gather sequences from GenBank IDs\n",
    "\n",
    "Author: Shane Gilligan-Steinberg\n",
    "Date: 240216\n",
    "\n",
    "Description:\n",
    "This script reads a list of GenBank IDs from a CSV file, fetches the corresponding nucleotide sequences \n",
    "from the GenBank database, and writes the sequences to a FASTA file. The GenBank IDs should be listed \n",
    "in a column named \"GenBank ID\" in the CSV file.\n",
    "This could be used to use a specific list of sequences for analysis. You will have to make:\n",
    "3. .csv list of subtypes (can use script 230704_Split_FASTA_v2_CRF.py to perform separation from LANL alignmnents) [~/Targets]\n",
    "- This will be needed to run this in the main script\n",
    "\n",
    "Additional Information:\n",
    "- Make sure to change the `csv_file` name in the script to match the path to your input CSV file.\n",
    "- The email address specified in the `Entrez.email` field should be updated to your email address for proper usage of the Entrez API.\n",
    "- Ensure that the CSV file contains a column named \"GenBank ID\" with the GenBank IDs you wish to fetch.\n",
    "- The output FASTA file will be saved as \"output.fasta\" in the same directory as the script.\n",
    "\"\"\"\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "def fetch_sequences(genbank_ids):\n",
    "    \"\"\"\n",
    "    Fetch sequences from GenBank given a list of GenBank IDs.\n",
    "    Args:\n",
    "        genbank_ids (list): List of GenBank IDs.\n",
    "    Returns:\n",
    "        list: List of SeqRecord objects.\n",
    "    \"\"\"\n",
    "    sequences = []\n",
    "    Entrez.email = \"\"  # Provide your email for Entrez\n",
    "\n",
    "    for genbank_id in genbank_ids:\n",
    "        try:\n",
    "            handle = Entrez.efetch(db=\"nucleotide\", id=genbank_id, rettype=\"gb\", retmode=\"text\")\n",
    "            record = SeqIO.read(handle, \"genbank\")\n",
    "            sequences.append(record)\n",
    "        except Exception as e:\n",
    "            print(f\"Error fetching sequence for {genbank_id}: {e}\")\n",
    "\n",
    "    return sequences\n",
    "\n",
    "def write_fasta(sequences, output_filename):\n",
    "    \"\"\"\n",
    "    Write sequences to a FASTA file.\n",
    "    Args:\n",
    "        sequences (list): List of SeqRecord objects.\n",
    "        output_filename (str): Output FASTA file name.\n",
    "    \"\"\"\n",
    "    with open(output_filename, \"w\") as fasta_file:\n",
    "        for sequence in sequences:\n",
    "            fasta_file.write(f\">{sequence.id}\\n{sequence.seq}\\n\")\n",
    "\n",
    "# Read CSV file with GenBank IDs\n",
    "csv_file = \"Test_GenBank.csv\"\n",
    "df = pd.read_csv(csv_file)\n",
    "print(df)\n",
    "\n",
    "# Extract GenBank IDs from the CSV DataFrame\n",
    "genbank_ids = df[\"GenBank ID\"].tolist()\n",
    "print(genbank_ids)\n",
    "\n",
    "# Fetch sequences from GenBank\n",
    "print(\"Fetching from GenBank........\")\n",
    "sequences = fetch_sequences(genbank_ids)\n",
    "\n",
    "# Write sequences to FASTA file\n",
    "output_fasta_file = \"output.fasta\"\n",
    "write_fasta(sequences, output_fasta_file)\n",
    "\n",
    "print(\"FASTA file generated successfully.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
